### ISSUE TITLE
xish' EventDispatcher grinds to a halt when using many different one-time observers

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11397]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)| @ralphm reported|
|-|-|
|Trac ID|trac#1652|
|Type|defect|
|Created|2006-04-19 12:13:42Z|

In `EventDispatcher`, (xpath query) observers are stored using a dictionary. The dictionary is keyed by a tuple of (query, priority), where query is a `XPathQuery` object. The value is a `CallbackList` object which holds the actual callback functions of the observers.

For one-time observers, when the callback has been called once, it is removed from the list of callbacks of the `CallbackList` object. However, even if there are no callbacks left, the entry in the dictionary of observers is not cleaned out.

When using `t.w.p.j.client.IQ`, each IQ request will add a new observer to the `EventDispatcher` of the stream to track answers by id. As each IQ has a unique id, this means a unique XPath query for every IQ request. As result the dictionary of `EventDispatchers` grows and grows, with references to objects that will never be used again, slowing things down tremendously.

Possible solutions:

  * Remove the entry in the dictionary when there are no observers for a particular query
    anymore.
  * Reimplement IQ to return deferreds on send, and having a generic IQ reply handler that fires
    the deferreds upon receipt of an IQ.

Attachments:

* [utility.py.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/ca7/ca70ca5f6efd8e48f2b4b75b3bd75a5f8fdb12d5/0950804f2c920d91b02b527c2c8db20bce40e118.patch) (493 bytes) - added by ralphm on 2006-04-19 12:18:39Z - Possible fix to cleanup observer dictionary when no observers are left for a particular xpath query
* [words05.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/ca7/ca70ca5f6efd8e48f2b4b75b3bd75a5f8fdb12d5/2175528490c9879b9ea0a71272242bb1e69ee3fd.diff) (784 bytes) - added by jack on 2007-01-31 18:22:34Z - Patch to utility.py to delete empty callback lists.
* [client.py.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/ca7/ca70ca5f6efd8e48f2b4b75b3bd75a5f8fdb12d5/73e0a704ece70d299c7c7e4368cd1d91b7a6b98a.diff) (791 bytes) - added by jack on 2007-01-31 18:23:20Z - Patch to client.py to skip over missing callback lists.
* [utility.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/ca7/ca70ca5f6efd8e48f2b4b75b3bd75a5f8fdb12d5/b76ea63eada9a62a18c1be01837a0ea65f5d595e.diff) (516 bytes) - added by jack on 2007-01-31 18:24:28Z - Patch to utility.py to skip over missing callback lists.
* [client.py.2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/ca7/ca70ca5f6efd8e48f2b4b75b3bd75a5f8fdb12d5/e7f5a164da1a95b34b292d713e8f56d39bc9a52a.diff) (533 bytes) - added by jack on 2007-01-31 18:25:28Z - Patch to client.py to ensure that there are callbacks.

<details><summary>Searchable metadata</summary>

```
trac-id__1652 1652
type__defect defect
reporter__ralphm ralphm
priority__highest highest
milestone__Words_0_6 Words-0.6
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__words words
keywords__ 
time__1145448822000000 1145448822000000
changetime__1173625958000000 1173625958000000
version__ 
owner__ 
cc__ralphm cc__jack cc__exarkun cc__radix cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)|@ralphm set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="jack's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|jack commented|
|-|-|

We also ran into this at Chesspark.  Our patch is similar to yours, but we have a few minor additions.  First, client.IQ should only be setting onetime observers on get or set iqs, not on result and error.  Next, your patch is incomplete.  While you delete the callback list appropriately, you have forgot to update the keylist.  And finally, in client.IQ._resultEvent, a small patch is needed if the callbacks are blank.  This may no longer be needed with the previous two patches, but we had to use it before.  I've attached our three patches, one of which is a slightly different implementation of your original patch.

I've switched type from 'enhancement' to 'defect'.  This is clearly a giant bug for anyone using client.IQ in a long running process.

**Comment by twisted-trac** [comment]
|<img alt="jack's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|jack commented|
|-|-|

Note, the first client.py.diff description is incorrect.  It is to ensure that onetime observers don't get added except for get/set iqs.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

Thanks for your persistence on this, jack.  I'll probably leave it up to ralphm to deal with this issue, but I will point out that it will be easier to fix this once there is test coverage for the code/bug in question.  If you want to push this forward as quickly as possibly, supplying that would be advantageous.  If you can't, that's fine too, someone else will take care of it, but it may take longer.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_5"></a>|@ralphm set owner to @exarkun<br>@ralphm set status to `new`|
|-|-|

I think I've come up with a reasonable test and fix for this problem. Unfortunately, since it is an internal administration issue, the only way I saw to test this was using private attributes of `EventDispatcher`.

Upon seeing this code again, it is a likely target for major refurbishment at a later time. I'll file tickets for them appropriately.

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_6"></a>|@ralphm set owner to @radix|
|-|-|

Found a regression and couldn't wait for refurbishment later on. Also split out the tests as suggested by radix.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve set owner to @ralphm|
|-|-|

A few things:
 * unused imports in test_xishutil
 * new tests methods should be name test_mixedCase
 * unused import in xish.utility
 * _orderedEventObserverKeys and _orderedXpathObserverKeys in EventDispatcher seem to be not used anymore
 * The 'observers = None' before call to self._getEventAndObservers seems useless too
 * types.StringTypes can simplify _isStr
 * I prefer a[:x] to a[0:x]
 * 'object' as a variable name is bad
 * in the docs, I think we must use C{int}/C{str} instead of L{int}/L{str}
 * += and -= are great :)

Overall, that seems good, with great docs and test coverage. I don't really apprehend the changes, because I don't know xish, but I guess that's ok.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)|@ralphm set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve set owner to @ralphm|
|-|-|

Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_10"></a>|@ralphm set status to `closed`|
|-|-|

(In [19669]) Cleanup empty callback lists in EventDispatcher.

Author: ralphm
Reviewer: therve
Fixes [#11397](https://github.com/twisted/twisted/issues/11397)

EventDispatcher did not remove empty callback lists in the event callback
registration. This causes events to be matched to queries that will not
trigger any callback being called, causing significant slowdown when
many one-time callbacks are registered.

This change makes sure empty callback lists are cleaned out, and also optimizes
the internal bookkeeping to not have to use sorting on every callback
registration. Instead the observers are kept in different dictionaries (query
-> callback list) for each priority level. Then, upon event dispatch, the
observers priorities are sorted and then each associated dictionary is
processed in that order. This assumes the number of used priority levels is
very low to be efficient.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_11"></a>|@ralphm set status to `reopened`|
|-|-|

Reverted r19669, because of the regression reported in [#2476](https://github.com/twisted/twisted/issues/2476).

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_12"></a>|@ralphm commented|
|-|-|

(In [19705]) Rebranch for reverted r19669 re [#11397](https://github.com/twisted/twisted/issues/11397) and [#2476](https://github.com/twisted/twisted/issues/2476).

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_13"></a>|@ralphm set owner to @exarkun<br>@ralphm set status to `new`|
|-|-|

The previous rebranch misfired, but this problem is now solved in [branches/xish-dispatcher-1652-2.](branches/xish-dispatcher-1652-2.) 

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set owner to @ralphm|
|-|-|

twisted.words.test.test_xishutil.EventDispatcherTest.test_addObserverTwice fails in the branch, and is missing a docstring.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_15"></a>|@ralphm set owner to @exarkun|
|-|-|

Ok, I'm stupidN. Note to self: never fix bugs in presentations. Instead, go out for beer and really fix bugs afterwards.

Please review again.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_16"></a>|@ralphm commented|
|-|-|

To ease the review process, you can find the diff with respect to the reverted merge [here](http://twistedmatrix.com/trac/changeset?old_path=%2Ftrunk&old=19669&new_path=%2Fbranches%2Fxish-dispatcher-1652-2&new=19715).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun set owner to @ralphm<br>@exarkun set status to `new`|
|-|-|

There's some trailing whitespace in utility.py and test_xishutil.py

The refactored line ```event = xpath.internQuery(event)``` which ended up on line 220 in this branch doesn't ever get invoked by the test suite.  If you'd rather fix this as part of a separate ticket, that's fine.

A lot of code dealing with priorities is being refactored, but I can't see any test coverage at all for the priority feature.  This should probably be fixed in this branch.

The change to fix the bug for which this branch was initially reverted looks good, though.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_19"></a>|@ralphm commented|
|-|-|

Fixed the whitespace.

Line 220 does not contain the line you mentioned. The only place that does have that code, does have coverage as I tested with `trial --coverage`.

The priority feature is at least tested by `testOrderedXPathDispatch`, maybe you missed that one.

Merging.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_20"></a>|@ralphm set status to `closed`|
|-|-|

(In [19807]) Cleanup empty callback lists in EventDispatcher.

Author: ralphm
Reviewer: therve, exarkun
Fixes [#11397](https://github.com/twisted/twisted/issues/11397), [[#2476](https://github.com/twisted/twisted/issues/2476)](https://github.com/twisted/twisted/issues/2476).

EventDispatcher did not remove empty callback lists in the event callback
registration. This causes events to be matched to queries that will not
trigger any callback being called, causing significant slowdown when
many one-time callbacks are registered.

This change makes sure empty callback lists are cleaned out, and also optimizes
the internal bookkeeping to not have to use sorting on every callback
registration. Instead the observers are kept in different dictionaries (query
-> callback list) for each priority level. Then, upon event dispatch, the
observers priorities are sorted and then each associated dictionary is
processed in that order. This assumes the number of used priority levels is
very low to be efficient.

This remerge also fixes a regression introduced in r19669 that made multiple
observers for the same query fail (#2476).

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
